stake-pool/single-pool/src/processor.rs

```diff
@@ -0,0 +1,1078 @@
+            .minimum_balance(stake_space)
+            .saturating_add(minimum_delegation);
+
+        // XXX TODO FIXME jon raised the important issue that if we check != someone can dos pool creation
+        // unfortunately, that complicates the token minting, because we undermint if someone overpays
+        // i believe we should be able to load the stake account data after delegation to see the true stake amount
+        // TODO also write a test that overpayment goes smoothly
+        if pool_stake_info.lamports() < stake_rent_plus_initial {
+            return Err(SinglePoolError::WrongRentAmount.into());
+        }
+
+        let authorized = stake::state::Authorized::auto(pool_authority_info.key);
+
+        invoke_signed(
+            &system_instruction::allocate(pool_stake_info.key, stake_space as u64),
+            &[pool_stake_info.clone()],
+            stake_signers,
+        )?;
+
+        invoke_signed(
+            &system_instruction::assign(pool_stake_info.key, stake_program_info.key),
+            &[pool_stake_info.clone()],
+            stake_signers,
+        )?;
+
+        invoke_signed(
+            &stake::instruction::initialize_checked(pool_stake_info.key, &authorized),
+            &[
+                pool_stake_info.clone(),
+                rent_info.clone(),
+                pool_authority_info.clone(),
+                pool_authority_info.clone(),
+            ],
+            authority_signers,
+        )?;
+
+        // delegate stake so it activates
+        invoke_signed(
+            &stake::instruction::delegate_stake(
+                pool_stake_info.key,
+                pool_authority_info.key,
+                validator_vote_info.key,
+            ),
+            &[
+                pool_stake_info.clone(),
+                validator_vote_info.clone(),
+                clock_info.clone(),
+                stake_history_info.clone(),
+                stake_config_info.clone(),
+                pool_authority_info.clone(),
+            ],
+            authority_signers,
+        )?;
+
+        // mint tokens to the user corresponding to their stake deposit
+        // note that this is safe from economic attacks despite the stake being in an activating state
+        // because we block anyone else from depositing until it has activated
+        // XXX actually do we give a shit about the activating stake?
+        // theres no economic attack anyway because theres no rewards! right??
+        // TODO reread this part of the stake program (or just test for ourselves actually...) that uh
+        // no weird shit like "merge into activating and then split" can implicitly deactivate...
+        Self::token_mint_to(
+            validator_vote_info.key,
+            token_program_info.clone(),
+            pool_mint_info.clone(),
+            user_token_account_info.clone(),
+            pool_authority_info.clone(),
+            authority_bump_seed,
+            minimum_delegation,
```

Comment:
I don't think this should mint even `minimum_delegation`, since that amount is untouchable by depositors. It can do `pool_stake_info.lamports() - stake_rent_plus_initial` however, since those lamports are usable.

Otherwise, someone could have pool tokens that they can never withdraw because of the minimums

Reply:
i kinda have been going back and forth on this. its not really that "those" tokens cant withdraw "their" stake, but rather that this part of the cost of initializing the pool is socialized among all depositors. but otoh 1sol isnt a lot to pay to make the pool, but minting the tokens slightly disincentivizes early deposits...

i think im going to not mint, but instead of messing with token supply, i realized i can just introduce "available stake" and the math should work the same